python/pyarrow/io.pxi

```diff
@@ -262,6 +271,158 @@ cdef class NativeFile:
         Alias for read, needed to match the IOBase interface."""
         return self.read(nbytes=None)
 
+    def readall(self):
+        return self.read()
+
+    def readinto(self, b):
+        """
+        Read into the supplied buffer
+
+        Parameters
+        -----------
+        b: any python object supporting buffer interface
+
+        Returns
+        --------
+        number of bytes written
+        """
+
+        cdef:
+            int64_t bytes_read
+            uint8_t* buf
+
+        self._assert_readable()
+
+        cdef Buffer py_buf = py_buffer(b)
+        cdef int64_t buf_len = py_buf.size
+
+        buf = py_buf.buffer.get().mutable_data()
+
+        with nogil:
+            check_status(self.rd_file.get().Read(buf_len, &bytes_read, buf))
+
+        return bytes_read
+
+    def readline(self, size=None):
+        """Read and return a line of bytes from the file.
+
+        If size is specified, read at most size bytes.
+
+        Line terminator is always b"\\n".
+
+        TODO: Move this method to libarrow."""
+
+        cdef:
+            int64_t c_nbytes
+            int64_t bytes_to_read
+            int64_t bytes_read
+            int64_t linebreak_pos = 0
+            int64_t initial_pos
+            uint8_t* linebreak_ptr
+            PyObject* obj
+            shared_ptr[CBuffer] read_buffer
+            int64_t READ_BUFFER_SIZE = 4096
+
+        self._assert_readable()
+
+        if size is None:
+            c_nbytes = self.size() - self.tell()
+        else:
+            c_nbytes = size
+
+        # ReadAt moves position inconsistently, so save the initial position
+        check_status(self.rd_file.get().Tell(&initial_pos))
+
+        with nogil:
+            while linebreak_pos < c_nbytes:
+                if self.rd_file.get().supports_zero_copy():
+                    # we try to read the whole file, since this operation
+                    # is zero copy
+                    check_status(self.rd_file.get().GetSize(&bytes_to_read))
+                else:
+                    # we read only READ_BUFFER_SIZE next bytes which might
+                    # not be the ideal size for every file type
+                    bytes_to_read = READ_BUFFER_SIZE
```

Comment:
A default 4kB buffer for reading a single line is definitely not the right size... Also I'm not sure it makes sense to destroy and reallocate a new buffer every time.

Reply:
Also there's no need to read that much if an explicit `size` was passed.